wgpu-hal/src/dx12/shader_compilation.rs

```diff
@@ -32,13 +32,21 @@ pub(super) fn compile_fxc(
     {
         compile_flags |= d3dcompiler::D3DCOMPILE_DEBUG | d3dcompiler::D3DCOMPILE_SKIP_OPTIMIZATION;
     }
+
+    // If no name has been set, D3DCompile wants the null pointer.
+    let source_name = if source_name.is_empty() {
+        ptr::null()
+    } else {
+        source_name.as_ptr()
+    };
```

Comment:
This is still unsound though: `source_name` doesn't have a trailing 0. If D3DCompile expects a null-terminated string, then allocation is required here.

Reply:
> This is still unsound though: `source_name` doesn't have a trailing 0. If D3DCompile expects a null-terminated string, then allocation is required here.

Hmm, the source is a `CString` if it is not `&""`, so `D3DCompile` won't know the difference, and `DxcCompiler::compile` reallocates. I can of course alter the signature of `compile_fxc` and `compile_dxc`, but that seemed like a more invasive change.

Reply:
I do agree it is Morally Wrong, of course.

Reply:
I deliberately used the term "unsound" -- the function is safe so it can't make assumptions about the callers, such as "this is secretly a CString". Also, if the caller guarantees that this is a CString, then your PR is unnecessary as a CString will always allocate at least one byte for the null terminator.

Also, under Stacked Borrows, if you get an `&str` then accessing the bytes outside the string length is UB.

Reply:
> Also, if the caller guarantees that this is a CString, then your PR is unnecessary as a CString will always allocate at least one byte for the null terminator.

It is an `Option<CString>, just to make matters confusing!

Reply:
Oh, the `&""` is hidden inside `unwrap_or_default`? I was wondering what the PR message talks about as there's no `&""` to be found in the code...

Reply:
YEP.

Reply:
I am slowly becoming of the opinion that `unwrap_or_default` was a mistake.

Reply:
Without knowing any of the context, this looks right to me. All I can really check is whether this is the proper way to convert `Option<&CStr>` to a C pointer.

Reply:
Using `&str` for the argument is also fine, but then it must allocate. As https://github.com/rust-lang/rust/issues/126442 shows, relying on the caller to uphold extra safety requirements (that are not documented!) is way too fragile.